v1.0.0-rc.1

ahmadnassri opened the pull request:
This is essentially a complete re-write and expansion on the original functionality of `v0.1.0`

- modular schema support, allows for easy future expansions
- support `RFC7239` & `X-Forwarded-*` as primary trusted schemas
- allows for implementers to choose schemas to process by passing an `options` object
- this *is* a breaking api change, where `v0.1.0` returns an array of IPs, `v1.0.0` returns an object representing parameters of `RFC 7239`
  - the version increment is required to allow dependants libraries/implementers to continue to operate safely with out version breakage *(assuming version resolving occurs with `npm`)
  - [`RFC 7239` Parameters](http://tools.ietf.org/html/rfc7239#section-5)
- coupled with support for less common schemas, named in association with vendor usage:
  - cloudflare
  - fastly
  - microsoft
  - nginx
  - zscaler

ahmadnassri committed:
v1.0.0-rc.1

This is essentially a complete re-write and expansion on the original functionality of `v0.1.0`

- modular schema support, allows for easy future expansions
- support `RFC7239` & `X-Forwarded-*` as primary trusted schemas
- allows for implementers to choose schemas to process by passing an `options` object
- this *is* a breaking api change, where `v0.1.0` returns an array of IPs, `v1.0.0` returns an object representing parameters of `RFC 7239`
  - the version increment is required to allow dependants libraries/implementers to continue to operate safely with out version breakage *(assuming version resolving occurs with `npm`)
  - [`RFC 7239` Parameters](http://tools.ietf.org/html/rfc7239#section-5)
- coupled with support for less common schemas, named in association with vendor usage:
  - cloudflare
  - fastly
  - microsoft
  - nginx
  - zscaler

ahmadnassri committed:
resolving node <=0.8 issue ("debug" target not found)

ahmadnassri committed:
add node 0.12 to travis matrix

ahmadnassri committed:
use node 0.6 safe versions of mocha + istanbul

ahmadnassri committed:
remove dependency on util._extend (node 0.6 support)

ahmadnassri committed:
fix gh markdown rendering bug

dougwilson commented on index.js:
```diff
@@ -1,35 +1,65 @@
+    host: req.headers && req.headers.host ? req.headers.host : undefined,
+    port: req.connection.remotePort.toString(),
+    ports: [req.connection.remotePort.toString()],
+    proto: req.connection.encrypted ? 'https' : 'http'
```
Are you sure this actually works in Node.js 0.6 and the others? I only ask because the test is just a mock, so we have to verify manually if we cannot add a real test.

ahmadnassri replied:
I was relying on the travis test to be honest, I managed to manually test all the way down to 0.8, then gave up on getting 0.6 installed on my machine. I'll get that working today and confirm.

dougwilson replied:
They are only as honest as your tests are--since you are using mocks, then you are not actually testing anything (you have to test a _real_ `req` for Travis CI to be of any help to you).

dougwilson commented on index.js:
```diff
@@ -1,35 +1,65 @@
-/*!
- * forwarded
- * Copyright(c) 2014 Douglas Christopher Wilson
- * MIT Licensed
```
We want to keep these headers so we know where in the source the copyrights lives (and so the note is retained when people copy the file but not the license).

ahmadnassri replied:
oops, forgot to re-add those! *(mind tends to zone those out!) :)*

dougwilson commented on index.js:
```diff
@@ -1,35 +1,65 @@
+  var opts = {
+    // default to only common + standard
+    // array order matters here
+    schemas: options.schemas || [
```
imo it should only default to a single schema, not multiple. The reason is people are mainly going to be set up for one.

ahmadnassri replied:
I would argue that typically in libraries such as this, it should be the standard only (i.e. `RFC 7239`).

however due to the long history of `X-Forwarded-*` and the relatively new age of `RFC 7239` it would make sense to consider both a *standard*.

in that light, both are the defacto default here, other scenarios would be:

- `RFC 7239` only: would essentially break every dependant's assumptions
- `X-Forwarded-*` only: would not be true to the standard and might even be problematic in the future.

this is precisely why the logic at [index.js#L52](https://github.com/jshttp/forwarded/blob/3e90938bf9fd143a995171e239dcd27d172c9af9/index.js#L52) overwrites properties in the sorted order of this schemas array, existing implementations of `X-Forwarded-*` will still operate identical to the past, and when `Forwarded` header is present, it counts as the authoritative source of information *as it is the standard*.

dougwilson replied:
> it counts as the authoritative source of information as it is the standard.

Yes, but this is the _biggest issue I have with this assumption_: people literally just are not using that header, nor are they blocking it. Because of this, it is extremely easy to fall in a trap where people rely on this module because customers are paying for IP whitelisting. An attacker can then just include a `Forwarded` header and even those the company's proxies are making sure `X-Forwarded-For` is correct, they will not have a breach of contract with the customer, as an attacker will include a `Forwarded` header and suddenly it becomes the authoritative source.

dougwilson replied:
Basically I think the standard should be the default and people can always change it (and we can simply add messaging to the README). I just don't like the default being multiple types, because I can just see the confusion/issues coming (I have a lot of experience with the things people report ;) ). I like the idea of the module letting you use multiple at once, though, if the user chooses.

ahmadnassri replied:
to clarify are you suggesting to keep the current RFC standard as the default option?

also, another option is to ***not offer a default**, leaving it to the implementer to decide what to support.

e.g. 

- `proxy-addr v1.0.8` will be `['xff']` and in next major release can switch to `['xff', 'rfc7239']`
- `express v4.x` in addressing (strongloop/express/pull/2666)  it could be `['xff']` only, next major release can be `['xff', 'rfc7239']` + with addition of a application setting for users to customize.

dougwilson replied:
> to clarify are you suggesting to keep RFC 7239 as the default option?

Sure that works. Specifically, I'm suggesting that the default option should be _only one schema_; which one it is really doesn't matter as much, because people will change it.

> another possible option is to not offer a default

Well, that is a possibility, yes, but I'm undecided on that (though if you did it, you'd need to move the `schemas` out of the `options` argument, since it's no longer optional).

dougwilson replied:
As for what `proxy-addr` or `express` may do, that's undecided for next majors, though of course for the existing major, they would stick to `'xff''`.

ahmadnassri replied:
resolved in 8694d1c

dougwilson commented on index.js:
```diff
@@ -1,35 +1,65 @@
+  }
+
+  // consistent case
+  opts.schemas.map(Function.prototype.call, String.prototype.toLowerCase)
```
Result of map is not being used. Either a mistake or should be forEach.

dougwilson commented on index.js:
```diff
@@ -1,35 +1,65 @@
+  return opts.schemas
+    // check if schemas exist
+    .filter(function (name) {
+      return schemas[name]
```
Unknown schemas need to throw an error, not be silently dropped. A user making a typo in a schema name may not realize for a while.

dougwilson commented on package.json:
```diff
@@ -1,33 +1,59 @@
   "devDependencies": {
-    "istanbul": "0.3.2",
-    "mocha": "~1.21.4"
+    "debug": "~2.2.0",
```
`debug` should be under `dependencies`, rather than `devDependencies`

dougwilson commented on index.js:
```diff
@@ -1,35 +1,65 @@
+    addrs: [req.connection.remoteAddress],
+    by: null,
+    host: req.headers && req.headers.host ? req.headers.host : undefined,
+    port: req.connection.remotePort.toString(),
```
On a disconnected socket, `req.connection.remotePort` is `undefined` and this will throw. This needs to be handled.

dougwilson commented:
>  proxy-addr works (with minor fix to its own test) 

it's actually a bug in your PR here that it relieved. the issue is that you are using a simple mock for `req`, but it does not fully test it. `req.connection.remotePort` can legitimately be `undefined`.

Here is the follow-though using Node.js 0.10.24:

1. Access to the `remotePort` getter at https://github.com/joyent/node/blob/v0.10.24/lib/net.js#L581
2. Call `this._getpeername()` https://github.com/joyent/node/blob/v0.10.24/lib/net.js#L561
3. Two places in the method return `{}` (https://github.com/joyent/node/blob/v0.10.24/lib/net.js#L563 and https://github.com/joyent/node/blob/v0.10.24/lib/net.js#L569).
4. Return the `port` property from `this._getpeername()`; the two cases from \#3 will thus return `undefined` for `remotePort`

dougwilson commented:
Another concern I have is why does this PR seem to be significantly slower? So this module does not have a benchmark suite, but `proxy-addr` does, so simply installing this PR into `proxy-addr` and changing the usage to `forwarded(req).addrs` results in the following difference:
```
#### forwarded 0.1.0

> node benchmark\matching.js

  1 test completed.
  2 tests completed.
  3 tests completed.
  4 tests completed.
  5 tests completed.
  6 tests completed.

  trust none     x 864,390 ops/sec ±1.62% (175 runs sampled)
  trust all      x 909,210 ops/sec ±1.55% (176 runs sampled)
  trust single   x  18,817 ops/sec ±1.37% (171 runs sampled)
  trust first    x 949,981 ops/sec ±1.30% (176 runs sampled)
  trust subnet   x  18,371 ops/sec ±1.55% (173 runs sampled)
  trust multiple x  18,834 ops/sec ±1.44% (174 runs sampled)
```
```
### forwarded PR

> node benchmark\matching.js

  1 test completed.
  2 tests completed.
  3 tests completed.
  4 tests completed.
  5 tests completed.
  6 tests completed.

  trust none     x 54,219 ops/sec ±1.54% (180 runs sampled)
  trust all      x 52,660 ops/sec ±1.75% (175 runs sampled)
  trust single   x 13,887 ops/sec ±1.55% (177 runs sampled)
  trust first    x 53,196 ops/sec ±1.36% (177 runs sampled)
  trust subnet   x 13,665 ops/sec ±1.39% (175 runs sampled)
  trust multiple x 13,615 ops/sec ±1.51% (180 runs sampled)
```

The matching benchmark is the closest one that measures just the `forwarded` module performance (mainly the "trust none", "trust all", and "trust first", which have almost no overhead other than simply calling `forwarded`).

As you can see above, based on those three benchmarks, performance dropped significantly. Since this module would theoretically be called on every request to a web server, it has a direct impact on the maximum req/sec a single Node.js instance can serve.

I understand, of course, that the current implement was very simply, but I did not expect to see it drop all the way down to 55k ops/sec. My main question is: can we do better? What are the bottlenecks in this code? Can v8 optimize functions, or does it keep bailing out for some reason?

This note on performance is not technically a blocker, though, but just something I noted and think one of us should at least glance into :)

dougwilson commented on lib/schemas/cloudflare.js:
```diff
@@ -0,0 +1,19 @@
+function isSecure (req) {
+  try {
+    var cf = JSON.parse(req.headers['cf-visitor'])
+    return cf.scheme !== undefined
```
This is broken; according to https://support.cloudflare.com/hc/en-us/articles/200170536-How-do-I-redirect-HTTPS-traffic-with-Flexible-SSL-and-Apache- the value `{"scheme":"http"}` is valid and this routine would incorrectly mark the request as secure.

dougwilson commented on lib/schemas/fastly.js:
```diff
@@ -0,0 +1,11 @@
+'use strict'
+
+function isSecure (req) {
+  return ~['1', 'true'].indexOf(req.headers['fastly-ssl'])
```
It seems according to the docs at https://docs.fastly.com/guides/backend-servers/maintaining-separate-http-and-https-requests-to-backend-servers just the existence of the header with any value is enough to signal SSL. Is there a reason to check the value here?

ahmadnassri commented:
thanks for all the great feedback, I'll make the appropriate changes.

dougwilson commented:
> thanks for all the great feedback

Absolutely! I wanted to get your the feedback quickly :) I noticed the repo was last changed in 2014, so I just make a few mundane updates to `master`; it is up to you if you want to merge it in your branch, rebase your branch on top, or simply do nothing w.r.t those updates :)

Also, with the significant code, please feel free to also add your name for 2015 in the LICENSE file.

dougwilson commented on lib/schemas/rfc7239.js:
```diff
@@ -0,0 +1,54 @@
+    return forwarded
+  }
+
+  splitMap(header, ELEMENT_SEPARATOR, function parseElement (el) {
```
It seems like this will fail to parse the header `Forwarded: for="foo,bar"` properly, opening a vector for security issues when relying on the header for access control.

ahmadnassri replied:
from my reading of [`RFC 7239`](http://tools.ietf.org/html/rfc7239#section-6) the spec states that each `for` parameter represents a single address. unless I'm reading this wrong. can you link to the section that says otherwise?

dougwilson replied:
That is irrelevant; an incoming header can contain any values, valid or not. The current implementation will do the wrong thing and this is not good, as the only reason you are parsing these headers is for logging, auditing, or access controls.

Just think of a standard proxy that does not modify anything existing in the `Forwarded` header and simply adds it's own data to it; it will pass-through the bogus value that will create false parses.

dougwilson replied:
Basically, what I'm saying is that the current parse is actually _not RFC 7239 compliant_, as you are not correctly following the ABNF, which allows for `,` and `;` to appear within a `quoted-string` and the defined ABNF for the header is as follows:

```
Forwarded   = 1#forwarded-element
forwarded-element = [ forwarded-pair ] *( ";" [ forwarded-pair ] )
forwarded-pair = token "=" value
value          = token / quoted-string
token          = 1*tchar
tchar          = "!" / "#" / "$" / "%" / "&" / "'" / "*"
               / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
               / DIGIT / ALPHA
quoted-string  = DQUOTE *( qdtext / quoted-pair ) DQUOTE
qdtext         = HTAB / SP /%x21 / %x23-5B / %x5D-7E / obs-text
obs-text       = %x80-FF
```

The given header, `Forwarded: for="foo,bar"` is syntactically valid, yet this module will definitely parse it incorrectly.

Because the RFC allows extensions, perhaps your proxy decides it's going to include a custom field of the authenticated user's name. Uh, oh!
```
# this is in bash, so the "" is an escaped double quote (i.e. it's only a single double quote in the program)
$ node -pe "require('forwarded')({connection:{remotePort:0},headers:{forwarded:'user=""bad,for=_whitelisted_ip"";for=_blacklisted_ip'}})"
{ addrs: [ 'whitelisted_ip', 'blacklisted_ip' ],
  by: null,
  host: undefined,
  port: '0',
  ports: [ '0' ],
  proto: 'http' }
```

Even though the header is 100% syntactically valid, with a single forwarded entry of `{for: '_blacklisted_ip', user: 'bad,for=_whitelisted_ip'}`, we end up thinking that the proxy got the request from `_whitelisted_ip` incorrectly due to the invalid parsing this module did :(

And I'm not even a security expert; I can tell you that if we don't fix the parsing, it will most likely end up as a critical security bug.

ahmadnassri replied:
>  The current implementation will do the wrong thing

can you clarify what you mean by that?

the [token syntax](http://tools.ietf.org/html/rfc7230#section-3.2.6) does not actually allow a comma to be present: 

```
     token          = 1*tchar

     tchar          = "!" / "#" / "$" / "%" / "&" / "'" / "*"
                    / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
                    / DIGIT / ALPHA
                    ; any VCHAR, except delimiters
```

```
Delimiters are chosen from the set of US-ASCII visual characters not allowed in a token
(DQUOTE and "(),/:;<=>?@[\]{}").
```
**so, if anything, we should discard `for="foo,bar"` since its not a valid token**

###### verification vs. validation

further, imo, the purpose of this library is to *parse* the headers, ***not to verify values***, the resulting values should be filtered through a secondary layer if the implementer chooses to verify values, such as network addresses, host names, etc ...

which leaves us with the question: **should we add validation here?**

specifically: to validate *addresses*, *hostnames*, *ports*.

---

as a side, RFC 7239 node identifier syntax could also include [unknown](http://tools.ietf.org/html/rfc7239#section-6.2) and [obfuscated](http://tools.ietf.org/html/rfc7239#section-6.3) identifiers.

ahmadnassri committed:
restore copyright notice

ahmadnassri committed:
throw error on invalid schema name

ahmadnassri committed:
lowecase schema names + test

ahmadnassri committed:
debug should be under dependencies, rather than devDependencies

ahmadnassri committed:
On a disconnected socket, req.connection.remotePort is undefined

ahmadnassri committed:
fix: address ordering in rfc2739 processor, fix: ensure processed 'for' property is an array

ahmadnassri committed:
default to rfc7239 schema

ahmadnassri committed:
fix: cloudflare https detection

ahmadnassri committed:
fix: fastly ssl header existance counts as HTTPS

ahmadnassri committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	.travis.yml
	index.js
	package.json
	test/test.js

ahmadnassri committed:
add author to LICENSE

ahmadnassri commented:
> I noticed the repo was last changed in 2014, so I just make a few mundane updates to master; it is up to you if you want to merge it in your branch, rebase your branch on top, or simply do nothing w.r.t those updates :)

resolved in 1761f9d

> Also, with the significant code, please feel free to also add your name for 2015 in the LICENSE file.

resolved in  7a3cb41

ahmadnassri committed:
update author info

ahmadnassri committed:
added test with real http server and request

ahmadnassri committed:
node 0.6 lacks a callback function on server.close

ahmadnassri commented:
node `v0.6` successful tests with real server / client :tada:

ahmadnassri commented:
progress paused for a couple of days due to workload, still experimenting with bnf parsers and regex setup

lpinca commented:
I wrote https://github.com/lpinca/forwarded-parse to parse the `Forwarded` header trying to follow the ABNF. It throws an error if the syntax is not valid.
It needs to be tested a lot more and maybe improved, but if it might be of interest, I'm more than happy to give you full access to repo or move it under this organization.

ahmadnassri commented:
@lpinca that's great, i've been bogged down with work and this is the last remaining piece in this PR to fully test / solve (proper parsing of `Forwarded`) i'll review your module and perhaps merge it into `lib/schemas/rfc7239.js` with credits of course!

ahmadnassri commented:
oh my, I've let this one slip ... lets try to get this merged before its' year anniversary :stuck_out_tongue: 

I'll review the work thus far, and look into the notes from @dougwilson whatever additional improvements are needed will be added.

ahmadnassri committed:
chore(dependencies): update dependencies + replace debug for lighter debug-log

ahmadnassri committed:
chore(rename): better use of header prefixes to reflect schema naming

ahmadnassri committed:
refactor(proto): simplified logic for protocol detection

- should gain a minor performance boost with simpler logic for protocol detection
- refactor tests for consistency

ahmadnassri committed:
refactor(rfc7239): use the excellent forwarded-parse by @lpinca

forwarded-parse provides an excellent parsing solution for RFC 7239 "Forwarded" header.
however the spec also allows for values beyond the scope of this module (e.g. adding arbitrary fields)
therefor it's best if forwarded-parse remains as a separate module whilte utilized here.

ahmadnassri committed:
feat(defaults): default schemas should be rfc7239 and x-forwarded

ahmadnassri committed:
feat(schemas): add weblogic proxy support

ahmadnassri committed:
docs(readme): update docs with details on each schema

ahmadnassri commented:
I've added some minor performance improvements and switched to using @lpinca `forwarded-parse` for better ANBF compatibility *(though it seems to be very [strict](https://github.com/lpinca/forwarded-parse/issues/1))*

next steps: 

- [ ] add benchmark suite _(in separate repo)_

ahmadnassri commented:
@dougwilson any direction from you to the todo list above?

dougwilson commented:
> and switched to using @lpinca  forwarded-parse  for better ANBF compatibility

This part I don't understand. The entire purpose of this module is literally to do that operation, not depend on another module to do it. This module's name marched the name of the header for a reason...

dougwilson commented:
Basically I think there is a huge misunderstanding of the current separation of concerns between the modules here.

This module is meant only to be concerned with parsing/formatting of the HTTP "Forwarded" header (and I suppose the X headers that it unified)

The proxy-addr module is designed to determine the "proxied address" of a request, which could include configuring the source of proxy headers.

This PR seems to be heavily mixing those two together into this single library to me.

dougwilson commented:
A good litmus test in this organization is that if your code logically organizes into more than a single JavaScript file, that is multiple modules, not one. There are still some old, inherited modules that have a lib folder, but we are slowing working to get rid of that such that every module consists only of a single index.js file (for runtime code).

dougwilson commented:
Oh, I know I keep making comments but I keep thinking of new things to say :) I have not yet looked at the new changes, or even really looked though to remind myself of the existing changes, but I think all the features and stuff in here are good, but more though from us (and perhaps the technical committee that now oversees this organization) needs to be done around which changes go into which modules, what new modules should be created, etc.

ahmadnassri commented:
> This part I don't understand. The entire purpose of this module is literally to do that operation, not depend on another module to do it. This module's name marched the name of the header for a reason...

I suppose I did not expand on my thoughts here... mea culpa, let me expand.

- `forwarded-parse` seems to be doing a proper job of parsing the `Forwarded` header in accordance with the spec.

- @lpinca has volunteered donating his work to this organization, which is great.

- this original `forwarded` module focused on `X-Forwarded-For` header, and only concerned itself with the IP addresses, including the socket address for the request (though the latter part could be considered out of scope, if we're talking only about headers)

- the RFC 7239 spec, introduces `by`, `host`, `proto` in addition to `for` as **core** parameters. *(while also allowing for extensions)*

so my thoughts up this point:

- the parsing module if donated into the org can continue to live separately and focuses on the spec *(while allowing extensions)*

- this module should focus on those **core** parameters, and any fallback that might be derived from defacto standards (such as `X-Forwarded-*` and `X-Real-*` headers) or vendor extensions used in popular cloud services.

- the `proxy-addr` continues to focus on Proxy IP functionality, leveraging info from this module.

hence the separation and added dependency, mostly due to the extension parsing allowed by spec, as I'm a fan of separation of concerns.

ahmadnassri commented:
> Oh, I know I keep making comments but I keep thinking of new things to say :) 

all good, same here :)

> more though from us (and perhaps the technical committee that now oversees this organization) needs to be done around which changes go into which modules, what new modules should be created, etc.

as I was reviewing the original work (almost a year old now!) I started thinking of better way to break this down too, I'm no longer a fan of having the vendor headers included in the core of this module.

I believe now, it should only concern it self with the RFC 7239 Spec and `X-Forwarded-*` only (meaning we can merge the work from `forwarded-parse` directly here.

and the vendor specs can be additional, separate modules, that can be passed to this one for additional processing *(e.g. building the stack of IPs in a desired order across multiple specs)*

not sure what a good pattern for the latter though, or how it would propagate across higher dependencies for user-level control? *(e.g. `proxy-addr` > `connect` > framework)*

a simple pattern can perhaps just be an array of module names, or directly references functions (that in turn can be loaded from modules) that provide a common object `{ for, by, host, proto }` that can then be used to build a stacked result.

ahmadnassri commented:
now that I'm back at a desktop keyboard, here's an example of how my previous thoughts might shape out:

###### minimal / default
```js
const forwarded = require('forwarded') // includes RFC2739, X-Forwarded-*
let result = forwarded(request)
```

###### verbose
```js
const forwarded = require('forwarded') 
const rfc2739 = require('forwarded-rfc2739')
const xforwarded = require('forwarded-x-forwarded')
const cloudflare = require('forwarded-cloudflare')
const fastly = require('forwarded-fastly')

// custom order
let options = {
  schemas: [cloudflare, rfc7239, fastly, xforwarded]
}

let result = forwarded(request, options)
```

dougwilson commented:
I think `proxy-addr` should be receiving those options, and this module is the one that does the parsing for one of those "schemas": `Forwarded`.

ahmadnassri commented:
would `proxy-addr` also deal with `for`, `by`, `host`, `proto` properties and data beyond just the "proxy address"

dougwilson commented:
There is no reason why it cannot do that, since those are all basically part of the "proxied address". Ideally one should at least get enough information to construct the URL for the request, and just the hostname/IP is not enough information, because it could be on a non-default port, could be using HTTP or HTTPS, and more.

Typically if people just wanted to parse their headers, they would directly depend on the module that just does the header parsing they need. The `proxy-addr` module was created to glue the pure parsing modules together with trust chains and expose some kind of interface that modules like Express or koa could consume for what they need.

I think the long-term plans is that `proxy-addr` will be the module that will get all the address information: host, scheme/protocol, port, etc.

It used to also parse things like `X-Forwarded-For`, but that was doing too much, so it was split out into this module, with the desire to get parsing of the real `Forwarded` header instead, though that header doesn't seem to have taken off very much (yet), so there hasn't been any requests to Express/Koa to even get alternative headers as part of the core parsing, though it would be good to have.

Because we keep going back and forth on this discussion, I still have not even had time to lok at your new code or even refresh myself on the existing content of this pull request, sorry!

ahmadnassri commented:
> Because we keep going back and forth on this discussion, I still have not even had time to lok at your new code or even refresh myself on the existing content of this pull request, sorry!

no worries, I think your point about `proxy-addr` is along the same lines of what I'm [suggesting](https://github.com/jshttp/forwarded/pull/1#issuecomment-201993914) 

I think we're debating semantics of where the logic sits.

The logic used here can be abstracted to `proxy-addr` and multiple modules / packages representing different "schems" can be created, then `proxy-addr` glues them together as you said.

###### TL;DR 

I think we're saying the same thing. :)

---

I'll let you review the current code first, then we can discuss further.

greenkeeperio-bot committed:
chore(package): update mocha to version 2.5.1

https://greenkeeper.io/

ahmadnassri committed:
Merge pull request #3 from ahmadnassri/greenkeeper-mocha-2.5.1

Update mocha to version 2.5.1 🚀

greenkeeperio-bot committed:
chore(package): update mocha to version 2.5.3

https://greenkeeper.io/

ahmadnassri committed:
Merge pull request #5 from ahmadnassri/greenkeeper-mocha-2.5.3

Update mocha to version 2.5.3 🚀

greenkeeperio-bot committed:
chore(package): update istanbul to version 0.4.3

https://greenkeeper.io/

ahmadnassri committed:
Merge pull request #1 from ahmadnassri/greenkeeper-istanbul-0.4.3

Update istanbul to version 0.4.3 🚀

greenkeeperio-bot committed:
chore(package): update mocha to version 3.0.1

https://greenkeeper.io/

ahmadnassri committed:
Merge pull request #8 from ahmadnassri/greenkeeper-mocha-3.0.1

Update mocha to version 3.0.1 🚀

greenkeeperio-bot committed:
chore: drop support for Node.js 0.10

BREAKING CHANGE: This module no longer supports Node.js 0.10

ahmadnassri committed:
Merge pull request #15 from ahmadnassri/greenkeeper/remove-node-0.10

👻😱 Node.js 0.10 is unmaintained 😱👻

greenkeeper[bot] committed:
chore(package): update dependencies

https://greenkeeper.io/

ahmadnassri committed:
Merge pull request #16 from ahmadnassri/greenkeeper/update-all

Update dependencies to enable Greenkeeper 🌴

greenkeeper[bot] committed:
chore(package): update mocha to version 3.2.0

https://greenkeeper.io/

ahmadnassri committed:
Merge pull request #17 from ahmadnassri/greenkeeper/mocha-3.2.0

Update mocha to the latest version 🚀
